hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/CSMMetrics.java

```diff
@@ -61,26 +63,39 @@ public class CSMMetrics {
   private @Metric MutableCounterLong numDataCacheMiss;
   private @Metric MutableCounterLong numDataCacheHit;
   private @Metric MutableCounterLong numEvictedCacheCount;
+  private @Metric MutableCounterLong pendingApplyTransactions;
 
   private @Metric MutableRate applyTransactionNs;
   private @Metric MutableRate writeStateMachineDataNs;
+  private @Metric MutableRate untilApplyTransactionNs;
+  private @Metric MutableRate startTransactionCompleteNs;
 
-  public CSMMetrics() {
+  public CSMMetrics(RaftGroupId gid) {
+    this.gid = gid;
     int numCmdTypes = ContainerProtos.Type.values().length;
     this.opsLatencyMs = new MutableRate[numCmdTypes];
+    this.opsQueueingDelay = new MutableRate[numCmdTypes];
     this.registry = new MetricsRegistry(CSMMetrics.class.getSimpleName());
     for (int i = 0; i < numCmdTypes; i++) {
       opsLatencyMs[i] = registry.newRate(
           ContainerProtos.Type.forNumber(i + 1).toString() + "Ms",
           ContainerProtos.Type.forNumber(i + 1) + " op");
+      opsQueueingDelay[i] = registry.newRate(
+          "queueingDelay" + ContainerProtos.Type.forNumber(i + 1),
+          ContainerProtos.Type.forNumber(i + 1) + " op");
     }
   }
 
   public static CSMMetrics create(RaftGroupId gid) {
     MetricsSystem ms = DefaultMetricsSystem.instance();
     return ms.register(SOURCE_NAME + gid.toString(),
```

Comment:
Is the closed Ratis `Pipeline` can be removed form the Metrics system? Will a Datanode that runs for a long time and has multiple Pipeline closes have many expired and invalid indicators?

Reply:
CSMMetrics is unregistered when the corresponding ContainerStateMachine is closed.

Maybe we should make a singleton CSMMetrics for a DataNode.

Reply:
Understand.